### PR TITLE
Use compiler integrated async phase under -Xasync

### LINKED ISSUE #26: support await in pattern guards
If possible.

In the meantime, I'll prohibited them explicitly.

### LINKED ISSUE #157: Compiler warns about "pure expression does nothing in statement position"
If you have a future `foo: Future[Unit]` and do a `await(foo)` the compiler will give a warning:

```
Warning:(258, 16) a pure expression does nothing in statement position; you may be omitting necessary parentheses
          await(foo)
               ^
```

Edit: This does not happen in all `await`s on `Future[Unit]`s, as I initially assumed.  Below is the minimal code I could get that reproduces this:

```
def foo(foobaz: Future[Unit]) =
  async {
    if ("".isEmpty) {
      await(foobaz)
      0
    }
  }
```

### LINKED ISSUE #197: Heed deprecation warnings by on 2.12 (master) branch
Before #195, the master branch was cross built for Scala 2.11 and 2.12, which incurred a large number of deprecation warnings in the `scala.reflect`.

Now that we've dropped 2.11 support from that branch, we can and should heed these warnings.

### LINKED ISSUE #202: a pure expression does nothing in statement position
Similar to scala/scala-async/issues/74

Here's the code snippet:
```
import scala.async.Async._
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global


def doStuff(v: Any): Any = {
  v
}

def main() {
  async {
    val s = if (1 == 1) {
      await(Future{doStuff(true)})
      await(Future{doStuff(10)})
      s"true"
    } else {
      s"false"
    }
    println(s)
  }
}

main()
```

Assigning the futures to a val will not throw the warning.

### LINKED ISSUE #220: scala.reflect.internal.annotations.uncheckedBounds in compiler mirror not found
We tried to upgrade our project to 0.10.0, but all compilation now fails with 

```
scala.ScalaReflectionException: object scala.reflect.internal.annotations.uncheckedBounds in compiler mirror not found.
    at scala.reflect.internal.Mirrors$RootsBase.staticClass(Mirrors.scala:129)
    at scala.reflect.internal.Mirrors$RootsBase.staticClass(Mirrors.scala:29)
    at scala.async.internal.TransformUtils.scala$async$internal$TransformUtils$$UncheckedBoundsClass(TransformUtils.scala:393)
    at scala.async.internal.TransformUtils.scala$async$internal$TransformUtils$$UncheckedBoundsClass$(TransformUtils.scala:392)
    at scala.async.internal.AsyncMacro$$anon$1.scala$async$internal$TransformUtils$$UncheckedBoundsClass$lzycompute(AsyncMacro.scala:25)
    at scala.async.internal.AsyncMacro$$anon$1.scala$async$internal$TransformUtils$$UncheckedBoundsClass(AsyncMacro.scala:25)
    at scala.async.internal.TransformUtils.uncheckedBounds(TransformUtils.scala:396)
    at scala.async.internal.TransformUtils.uncheckedBounds$(TransformUtils.scala:394)
    at scala.async.internal.AsyncMacro$$anon$1.uncheckedBounds(AsyncMacro.scala:25)
    at scala.async.internal.AsyncTransform.asyncTransform(AsyncTransform.scala:30)
    at scala.async.internal.AsyncTransform.asyncTransform$(AsyncTransform.scala:24)
    at scala.async.internal.AsyncMacro$$anon$1.asyncTransform(AsyncMacro.scala:25)
    at scala.async.internal.AsyncBase.asyncImpl(AsyncBase.scala:56)
    at scala.async.internal.ScalaConcurrentAsync$.asyncImpl(ScalaConcurrentAsync.scala:27)
```

Scala version is 2.12.8 and sbt 1.2.8.

### PR DESCRIPTION
**Pull request by retronym**
See https://github.com/scala/scala/pull/8816

Most of the tests are deleted here as they have been migrated into scala/scala.

Fixes #26
Fixes #157
Fixes #197
Fixes #202
Fixes #220 (replacing the documented workaround)
Fixes #237 (2.12.12 is now included in this PR, we didn't do this in stages)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
+### Use a modern Scala compiler
+
+As of scala-async 1.0, Scala 2.12.12+ or 2.13.3+ are required.
+
+### Add dependency
+
+#### SBT Example
+
 ```scala
 libraryDependencies += "org.scala-lang.modules" %% "scala-async" % "0.10.0"
```
Comment (lrytz) [inline README.md:17]:
```suggestion
libraryDependencies += "org.scala-lang.modules" %% "scala-async" % "1.0.0"
```

File: README.md [hunk 2]
```diff
@@ -93,6 +127,22 @@ def combined: Future[Int] = async {
...
 
+## Limitations
+
+### `await` must be directly in the control flow of the async expression
+
+The `await` cannot be nested under a local method, object, class or lambda:  
+
+```
+async {
+  List(1).foreach { x => await(f(x) } // invali
```
Comment (lrytz) [inline README.md:138]:
```suggestion
  List(1).foreach { x => await(f(x) } // invalid
```

File: src/main/scala/scala/async/Async.scala [hunk 3]
```diff
@@ -50,14 +51,42 @@ object Async {
...
    * Non-blocking await the on result of `awaitable`. This may only be used directly within an enclosing `async` block.
    *
    * Internally, this will register the remainder of the code in enclosing `async` block as a callback
    * in the `onComplete` handler of `awaitable`, and will *not* block a thread.
    */
-  @compileTimeOnly("`await` must be enclosed in an `async` block")
+  @compileTimeOnly("[async] `await` must be enclosed in an `async` block")
   def await[T](awaitable: Future[T]): T = ??? // No implementation here, as calls to this are translated to `onComplete` by the macro.
+
+  def asyncImpl[T: c.WeakTypeTag](c: whitebox.Context)
```
Comment (lrytz) [inline src/main/scala/scala/async/Async.scala:65]: why `whitebox`?

File: src/main/scala/scala/async/Async.scala [hunk 4]
```diff
@@ -50,14 +51,42 @@ object Async {
...
+                                          (execContext: c.Tree): c.Tree = {
+    import c.universe._
+    if (!c.compilerSettings.contains("-Xasync")) {
+      c.abort(c.macroApplication.pos, "The async requires the compiler option -Xasync (supported only by Scala 2.12.12+ / 2.13.3+)")
+    } else try {
+      val awaitSym = typeOf[Async.type].decl(TermName("await"))
+      def mark(t: DefDef): Tree = {
+        import language.reflectiveCalls
+        c.internal.asInstanceOf[{
+          def markForAsyncTransform(owner: Symbol, method: DefDef, awaitSymbol: Symbol, config: Map[String, AnyRef]): DefDef
```
Comment (lrytz) [inline src/main/scala/scala/async/Async.scala:76]: this cast was not needed in the test implementation over at scala/scala, why is it here?

File: .travis.yml [hunk 5]
```diff
@@ -5,8 +5,9 @@ import: scala/scala-dev:travis/default.yml
 language: scala
 
 scala:
-  - 2.12.11
-  - 2.13.2
+  # Add when released
+  # - 2.12.12 when
```
Comment (SethTisue) [inline .travis.yml:9]: I'd suggest opening an issue to track this.

File: README.md [hunk 6]
```diff
@@ -119,53 +169,3 @@ The `async` approach has two advantages over the use of
...
-     called on completion of each future. The behavior of
-     this method is determined by the current state. It records
-     the downcast result of the future in a field, and calls the
-     `resume()` method.
-   - the `resume(): Unit` method that switches on the current state
-     and runs the users code for one 'chunk', and either:
-       a) registers the state machine as the handler for the next future
-       b) completes the result Promise of the `async` block, if at the terminal state.
-   - an `apply(): Unit` method that starts the computation.
-
```
Comment (SethTisue) [inline README.md:163]: Is the idea that this removed section will now  live somewhere else

File: src/main/scala/scala/async/Async.scala (hunk 4 above)
Comment (retronym) [reply src/main/scala/scala/async/Async.scala]: Removed, it was only needed in a prior version of this PR that cross-build with and without `-Xasync` support in the compiler. I've left in the `catch { case _: ReflectiveOperationException } =>` to give a better error message if `scala-reflect.jar` is older than `scala-compiler.jar` for some reason.

File: README.md (hunk 6 above)
Comment (retronym) [reply README.md:163]: https://github.com/scala/scala-async/pull/237

File: src/main/scala/scala/async/Async.scala (hunk 3 above)
Comment (retronym) [reply src/main/scala/scala/async/Async.scala]: Changed to blackbox.

File: .travis.yml (hunk 5 above)
Comment (retronym) [reply .travis.yml:9]: https://github.com/scala/scala-async/issues/241

File: README.md (hunk 6 above)
Comment (retronym) [reply README.md:163]: It is now: https://github.com/scala/scala-async/issues/242

**Comment by retronym** [comment]
Yep, 2.12.12 will be out within a week so we can wait.

**Comment by retronym** [comment]
Pushed the update to use the freshly minted 2.12.12

**Comment by retronym** [comment]
Cutting 1.0.0-M1 with this PR merged: https://github.com/scala/scala-async/releases/tag/v1.0.0-M1

File: README.md (hunk 1 above)
Comment (jsdt) [reply README.md:19]: Looks like this should be "1.0.0-M1"

**Pull request merged**